zebra-network/src/meta_addr.rs

```diff
@@ -160,135 +823,244 @@ impl MetaAddr {
-    ///     that sent us this address
-    ///   - `Failed`: the last time we marked the peer as failed
-    ///   - `AttemptPending`: the last time we queued the peer for a reconnection
-    ///     attempt
+    /// See [`PeerAddrState::get_last_success`] for details.
+    pub fn get_last_success(&self) -> Option<DateTime<Utc>> {
+        self.last_connection_state.get_last_success()
+    }
+
+    /// The last time a direct outbound connection to the address of this peer
+    /// failed.
+    ///
+    /// See [`PeerAddrState::get_last_failed`] for details.
+    pub fn get_last_failed(&self) -> Option<DateTime<Utc>> {
+        self.last_connection_state.get_last_failed()
+    }
+
+    /// The last time another node claimed this peer was valid.
+    ///
+    /// The exact meaning depends on [`last_connection_state`]:
+    ///   - [`NeverAttemptedSeed`]: the seed config doesn't have any last seen info,
+    ///      so this field is [`None`]
+    ///   - [`NeverAttemptedGossiped`]: the unverified time provided by the remote
+    ///      peer that sent us this address
+    ///   - [`NeverAttemptedAlternate`]: the local time we received the [`Version`]
+    ///      message containing this address from a peer
+    ///   - [`Failed`] and [`AttemptPending`]: these states do not update this field
     ///
     /// ## Security
     ///
-    /// `last_seen` times from `NeverAttempted` peers may be invalid due to
+    /// last seen times from non-`Responded` peers may be invalid due to
     /// clock skew, or buggy or malicious peers.
-    pub fn get_last_seen(&self) -> DateTime<Utc> {
-        self.last_seen
+    ///
+    /// Typically, this field should be ignored, unless the peer is in a
+    /// never attempted state.
+    pub fn get_untrusted_last_seen(&self) -> Option<DateTime<Utc>> {
+        self.last_connection_state.get_untrusted_last_seen()
+    }
+
+    /// The last time we successfully made a direct outbound connection to this
+    /// peer, or another node claimed this peer was valid.
+    ///
+    /// Clamped to a [`u32`] number of seconds.
+    ///
+    /// [`None`] if the address was supplied as a seed.
+    ///
+    /// ## Security
+    ///
+    /// last seen times from non-`Responded` peers may be invalid due to
+    /// clock skew, or buggy or malicious peers.
+    ///
+    /// Use [`get_last_success`] if you need a trusted, unclamped value.
+    pub fn get_last_success_or_untrusted(&self) -> Option<DateTime<Utc>> {
+        // Use the best time we have, if any
+        let seconds = self
+            .get_last_success()
+            .or_else(|| self.get_untrusted_last_seen())?
+            .timestamp();
+
+        // Convert to a DateTime
+        let seconds = seconds.clamp(u32::MIN.into(), u32::MAX.into());
+        let time = Utc
+            .timestamp_opt(seconds, 0)
+            .single()
+            .expect("unexpected invalid time: all u32 values should be valid");
+
+        Some(time)
+    }
```

Comment:
Security: Zebra should stop gossiping failure times as last_seen times #1868

Part 2: Get a last success time that can be serialized into an `Addr` message
* prefer our local last success time (which is updated on every message)
* if we don't have a local time, use the gossiped or alternate time
* clamp to a `u32` so it will fit in an `Addr` message